### ISSUE TITLE
Split VNNI and TPP

### ISSUE THREAD

**Issue by chelini** [plaidml/tpp-mlir#385]
VNNI currently converts to TPP. This means that in TPP we need to duplicate VNNI ops and there is not need for that since VNNI works on memref too, see:
https://github.com/plaidml/tpp-mlir/blob/61519866f8e23ffa10f5ee224af2418b82201608/include/TPP/Dialect/Tpp/TppOps.td#L299
and
https://github.com/plaidml/tpp-mlir/blob/main/include/TPP/Dialect/VNNI/VNNIOps.td#L59

I would like to split VNNI from TPP. The new flow would be:
VNNI -> XSMM
instead of
VNNI -> TPP -> XSMM

Comments: @KavithaTipturMadhu @adam-smnk and @rengolin

**Comment by adam-smnk** [comment]
Isn't VNNI more high-level or at least at the same abstraction level as TPP? If so, it should be possible to express all VNNI ops as existing TPP ops. Plus I think all of our passes work on TPP ops anyway so, we probably need this intermediate step to able to optimize IR further.
Or do you have something else in mind?

**Comment by chelini** [comment]
Yes, VNNI is at much higher level than TPP but works both at memref and tensor. Today, I see a bit of duplication: VNNI ops at memref level are simply translated 1:1 with TPP ops at memref level (https://github.com/plaidml/tpp-mlir/blob/61519866f8e23ffa10f5ee224af2418b82201608/lib/TPP/ConvertVNNIToTpp.cpp#L21). We need to maintain these TPP ops just to convert from VNNI. We could either restrict VNNI to be a tensor level dialect or drop the ops from TPP and decouple VNNI from it.

**Comment by chelini** [comment]
Actually the only way is to decouple the two.

**Comment by adam-smnk** [comment]
Right, I forgot there are these special `tpp.vnni_*` ops. Then indeed it is pointless duplication.

**Comment by chelini** [comment]
For VNNI to XSMM lowering we can reuse the TPP lowering, it is not a big change. I just don't like the fact of having duplicate ops.

**Comment by alheinecke** [comment]
BTW, the only reasons why there is tpp.vnnibrgemm etc, is because the tpp dialect doesn't handle flags well and a very finite set of flags is supported by pulling them into the function call. for tpp.vnnibrgemm : a brgemm where A matrix is VNNI layout....

**Comment by alheinecke** [comment]
> This could be another reason on tpp on tensor.

I would not go that far. VNNI-Layouts a dictated by hardware & necessary evil. For performance TPP bubble them up in pass-through fashion, however, even when mapping to larger GEMM, or conv operators, etc. VNNI layout (in some form, not necessarily as used in our code) is a must. So coupling VNNI & TPP wrt. to tensor can help us, but it's not a general solution for VNNI layout challenges. Most of the later chechs and pitfalls live really on buffer-level.

I see TPP tensor as extra work right now (which needs a lot of thinking and planning), as we still have to double check after bufferization that everything is in order, see #384 which right blocks mapping to TPP as bufferization messes it up. In case of #384 and having TPP on tensor, it would fuse on tensor level, bufferization messes it, up so the fused TPP needs to be undone to ensure correctness.

**Comment by rengolin** [comment]
> I see TPP tensor as extra work right now (which needs a lot of thinking and planning), as we still have to double check after bufferization that everything is in order, see #384 which right blocks mapping to TPP as bufferization messes it up. In case of #384 and having TPP on tensor, it would fuse on tensor level, bufferization messes it, up so the fused TPP needs to be undone to ensure correctness.

Agreed, TPP on tensors has more downsides than upsides. Any dialect that supports both needs considerations, which peppers the code with `hasBufferSemantics` and `hasTensorSemantics` and does not make things simpler (different number of operands, output control). If we can keep that complexity at `linalg` level and only handle memrefs at TPP level, it would be a lot simpler.

**Comment by rengolin** [comment]
We'll kill VNNI (#389), so this is moot.
